java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java

```diff
@@ -111,47 +173,32 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
 
     @Override
     public void visitMethodInvocation(MethodInvocationTree tree) {
-      if (EQUALS_MATCHER.matches(tree)) {
+      if (tree.methodSelect().is(Tree.Kind.IDENTIFIER) || isSuperOrThis(((MemberSelectExpressionTree) tree.methodSelect()).expression())) {
         typeChecked = true;
+      } else {
+        super.visitMethodInvocation(tree);
       }
     }
 
-    private boolean isArgument(ExpressionTree tree) {
-      ExpressionTree expressionTree = removeParenthesis(tree);
-      return expressionTree.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expressionTree).symbol().equals(parameterSymbol);
-    }
-
-    private boolean isExplicitComparison(ExpressionTree operand1, ExpressionTree operand2) {
-      return (isArgument(operand1) && isThis(operand2)) || (isArgument(operand2) && isThis(operand1));
-    }
-
-    private boolean isGetClassOnArgument(ExpressionTree tree) {
+    private boolean isSuperOrThis(ExpressionTree tree) {
```

Comment:
Could indeed happen in some kind of proxy object.

Changed so that a method call type checking if it takes the argument as input:
this.bar(that) // Compliant
this.foo.bar(null, that) // Compliant

this.bar() // Noncompliant, that is not referenced
this.foo.bar() // Noncompliant, that is not referenced